Bucket color by category

elenatorro opened the pull request:
Related issue https://github.com/CartoDB/carto-vl/issues/519

This PR covers the following cases:

**updated**

### when palettes are color arrays

#### and properties are categories

  - and there are less categories than colors
    - and not all categories in the dataset have a bucket defined
      - [x] should not show interpolation
      - [x] should use last color for the remaining categories
      - [x] should use only the same number of colors as categories

    - and all categories in the dataset have a bucket defined
      - [x] should show interpolation

  - and there are the same number of categories than colors
    - and not all categories in the dataset have a bucket defined
      - [x] should not show interpolation
      - [x] should use the last color for the last category
      - [x] should use last color for the remaining categories
    - and all categories in the dataset have a bucket defined
       - [x] should not show interpolation

  - and there are more categories than colors
    - and not all categories in the dataset have a bucket defined
        - [x] should use the default color for the remaining categories
  
#### and properties are numbers
   - and there are less buckets than colors 
    `ramp(buckets((1, [10, 20, 30]), [red, blue, yellow, green, purple])`
        - [x] should not show interpolation
        - [x] should ignore the remaining colors
   - and there are the same number of buckets than colors
      `ramp(buckets((1, [10, 20, 30]), [red, blue, yellow, green])`
        - [x] should not show interpolation
   - and there are more buckets than colors
       `ramp(buckets((1, [10, 20, 30, 40, 50, 60]), [red, blue, yellow, green])`
        - [x] should show interpolation

### when palettes are defined palettes

  - and palettes are qualitative
    - and not all categories in the dataset have a bucket defined
       - [x] should not show interpolation
       - [x] should use last color for the remaining categories
    - and all categories in the dataset have a bucket defined
       - [x] should not show interpolation
       - [x] should not set the last ramp color to the last category

  - and palettes are quantitative
    - and there are less or equal categories than the max quantitative ramp length (7)
      - [x] should not show interpolation
    - and there are more categories than the max quantitative ramp length (7)
      - [x] should show interpolation

IagoLast commented on test/unit/core/viz/expressions/ramp.test.js:
```diff
@@ -17,19 +20,374 @@ describe('src/core/viz/expressions/ramp', () => {
+            let expected;
+
+            describe('and there are less categories than colors', () => {
+                describe('with "others"', () => {
```
I find `and all categories in the dataset have a bucket defined` more descriptive ;)

elenatorro replied:
fixed :)

IagoLast commented on test/unit/core/viz/expressions/ramp.test.js:
```diff
@@ -17,19 +20,374 @@ describe('src/core/viz/expressions/ramp', () => {
+
+            describe('and there are less categories than colors', () => {
+                describe('with "others"', () => {
+                    it('should not show interpolation', () => {
```
`should return a non-interpolated color`

IagoLast replied:
ok, I just found all colors are interpolated so we need a better description here

IagoLast commented on test/unit/core/viz/expressions/ramp.test.js:
```diff
@@ -17,19 +20,374 @@ describe('src/core/viz/expressions/ramp', () => {
+                });
+
+                describe('without "others"', () => {
+                    it('should show interpolation', () => {
```
`should return an interpolated color`

IagoLast replied:
same

IagoLast commented on test/unit/core/viz/expressions/ramp.test.js:
```diff
@@ -17,19 +20,374 @@ describe('src/core/viz/expressions/ramp', () => {
+                        actual = r.eval();
+                        expected = yellow._nameToRGBA();
+        
+                        expect(actual).not.toEqual(expected);
```
This assertion is not strong enough, can we expect the interpolated color instead?

elenatorro replied:
I will think a better test for this case, but the answer is no, we can not expect the interpolated color. First, because we do not care about the precise interpolated value, and second, because it would be a weak test case, hard to understand and that we can break it easily if we change the way we interpolate.

IagoLast replied:
we can always use [toBeCloseTo](https://jasmine.github.io/api/2.6/matchers.html#toBeCloseTo) with a threshold

elenatorro replied:
I like that!

elenatorro commented:
Hi, @makella , thank you very much for this, I know there are a lot of cases. I am making some changes to the test descriptions as @IagoLast suggested, because they are hard to understand (specially with/without "others" section)

## Case 2:

In this case it means that we do not have 'others', we are using all the categories in the column. So it would be:

(*total categories: A, B, C, D*)
`ramp(buckets($cat,[A,B,C,D]),[red,orange,yellow,green, blue]`

Therefore, we have to **interpolate** the colors because there are more colors than values.

## Case 3

> there wouldn't be the need for the third item (use the default color for "others"), right?

Yes, exactly :)

> This would be the default for Case 3, right? Since categories = colors?

Yes :)

## Case 4

>  I don't think we need to consider whether or not there are "others", right?

That is true! I am removing this test.

> Do we need to worry about this for this case?

- `should use the last color for the last category`: True, this is not necessary.

- `should use the default color for others`: I was not sure about this case, cause I think it is still open. I am not sure if, by having less colors, we should:

(*total categories: A, B, C, D, E*)
`ramp(buckets($cat,[A,B,C,D]),[red,orange,blue]`
* a) use the default color for others
Interpolate red, orange and blue from A to D, and E would be **default**.

* b) use the last color for others
Interpolate red and orange from A to D, and E would be **blue**.

IagoLast commented:
my two cents about this case:


>total categories: A, B, C, D)
>ramp(buckets($cat,[A,B,C,D]),[red,orange,yellow,green, blue]

In this case since there are more colors than values:

- Throw an error
- Use the first n required colors and ignore the remaining ones

elenatorro commented:
That is totally right, @IagoLast. I think I already discussed that with @davidmanzanares one week ago, so I would go for:

> Use the first n required colors and ignore the remaining ones.

makella commented:
> total categories: A, B, C, D)
> ramp(buckets($cat,[A,B,C,D]),[red,orange,yellow,green, blue]
> Use the first n required colors and ignore the remaining ones

Agree!

makella commented:
@elenatorro as discussed, I'll test these scenarios locally today and report back!

makella commented:
@elenatorro @IagoLast 

For the cases we have defined above, I am going to test each one and document here. 

I think I captured all of scenarios for the categories/custom colors. The one case I have a question about is **2b**.

I haven't gone through the CARTOColor scenarios yet! I will do that tomorrow (doing a Builder workshop for the rest of the day today).

## Testing Data

I am using a Barcelona Barris dataset where the `n_distri` attribute has **10** unique districts:

`"Les Corts","Ciutat Vella","Sant Mart","Nou Barris","Eixample","Grcia","Horta-Guinard","Sarri-Sant Gervasi","Sant Andreu","Sants-Montjuc"`

Here is a map with the data. If you want to do the same tests for each case that I am, you can copy/paste the styling from each cases "Test" section.

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiIiLCJmIjp7ImxuZyI6Mi4xODc4NzY4ODc4MzY5NzgzLCJsYXQiOjQxLjQwMDc4NTI1NDY5NzkyfSwiZyI6MTEuNzEzNzIyMjg5MjQxMTc3LCJoIjoiRGFya01hdHRlciJ9

## Case 1a
A user wants to color four districts with a set of defined colors (`red,orange,yellow,green`) and have all other features colored with an "other" color that they have defined (`blue`).

### Test
`color: ramp(buckets($n_distri,["Les Corts","Ciutat Vella","Sant Mart","Nou Barris"]),[red,orange,yellow,green,blue])`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcChidWNrZXRzKCRuX2Rpc3RyaSxbXCJMZXMgQ29ydHNcIixcIkNpdXRhdCBWZWxsYVwiLFwiU2FudCBNYXJ0XCIsXCJOb3UgQmFycmlzXCJdKSxbcmVkLG9yYW5nZSx5ZWxsb3csZ3JlZW4sYmx1ZV0pIiwiZiI6eyJsbmciOjIuMjQxNDE2MTY4MjIxNTc2MywibGF0Ijo0MS4zODk1NDI2ODQ1NjAzfSwiZyI6MTAuOTcwNTY2NTU0MzAwNDU2LCJoIjoiRGFya01hdHRlciJ9

### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

## Case 1b
A user has defined an equal amount of category buckets to colors. In this case, we will color each bucket with the list of defined colors and color everything else with a default "other" color.

### Test
`color: ramp(buckets($n_distri,["Les Corts","Ciutat Vella","Sant Mart","Nou Barris"]),[red,orange,yellow,green])`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcChidWNrZXRzKCRuX2Rpc3RyaSxbXCJMZXMgQ29ydHNcIixcIkNpdXRhdCBWZWxsYVwiLFwiU2FudCBNYXJ0XCIsXCJOb3UgQmFycmlzXCJdKSxbcmVkLG9yYW5nZSx5ZWxsb3csZ3JlZW5dKSIsImYiOnsibG5nIjoyLjI0MTQxNjE2ODIyMTU3NjMsImxhdCI6NDEuMzg5NTQyNjg0NTYwM30sImciOjEwLjk3MDU2NjU1NDMwMDQ1NiwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [x] **Yes**
- [ ] No
- [ ] Unsure

## Case 2a
A user defines a bucket for each category in the attribute (in this case 10 districts), but adds an additional color (in this case `slategray`).

In this case, we will use all of the colors in the list and ignore the extra (in this case, 11th `slategray`) color.

### Test

`color: ramp(buckets($n_distri,["Les Corts","Ciutat Vella","Sant Mart","Nou Barris","Eixample","Grcia","Horta-Guinard","Sarri-Sant Gervasi","Sant Andreu","Sants-Montjuc"]),[red,orange,yellow,green,blue,purple,sienna,fuchsia,coral,turquoise,slategray])`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcChidWNrZXRzKCRuX2Rpc3RyaSxbXCJMZXMgQ29ydHNcIixcIkNpdXRhdCBWZWxsYVwiLFwiU2FudCBNYXJ0XCIsXCJOb3UgQmFycmlzXCIsXCJFaXhhbXBsZVwiLFwiR3JjaWFcIixcIkhvcnRhLUd1aW5hcmRcIixcIlNhcnJpLVNhbnQgR2VydmFzaVwiLFwiU2FudCBBbmRyZXVcIixcIlNhbnRzLU1vbnRqdWNcIl0pLFtyZWQsb3JhbmdlLHllbGxvdyxncmVlbixibHVlLHB1cnBsZSxzaWVubmEsZnVjaHNpYSxjb3JhbCx0dXJxdW9pc2Usc2xhdGVncmF5XSkiLCJmIjp7ImxuZyI6Mi4yNDE0MTYxNjgyMjE1NzYzLCJsYXQiOjQxLjM4OTU0MjY4NDU2MDN9LCJnIjoxMC45NzA1NjY1NTQzMDA0NTYsImgiOiJEYXJrTWF0dGVyIn0=
 
### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

## Case 2b

**_Needs further discussion_**

A user has defined a subset of district buckets (8) and has 11 colors defined. 

In this case, what I see happening is that we use the first color `red` for the first district `Les Corts` and the last color `slategray` for the two undefined category buckets `"Sant Andreu","Sants-Montjuc"`. All of the other districts are interpolated between the remaining colors.

### Test
`color: ramp(buckets($n_distri,["Les Corts","Ciutat Vella","Sant Mart","Nou Barris","Eixample","Grcia","Horta-Guinard","Sarri-Sant Gervasi"]),[red,orange,yellow,green,blue,purple,sienna,fuchsia,coral,turquoise,slategray])`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcChidWNrZXRzKCRuX2Rpc3RyaSxbXCJMZXMgQ29ydHNcIixcIkNpdXRhdCBWZWxsYVwiLFwiU2FudCBNYXJ0XCIsXCJOb3UgQmFycmlzXCIsXCJFaXhhbXBsZVwiLFwiR3JjaWFcIixcIkhvcnRhLUd1aW5hcmRcIixcIlNhcnJpLVNhbnQgR2VydmFzaVwiXSksW3JlZCxvcmFuZ2UseWVsbG93LGdyZWVuLGJsdWUscHVycGxlLHNpZW5uYSxmdWNoc2lhLGNvcmFsLHR1cnF1b2lzZSxzbGF0ZWdyYXldKSIsImYiOnsibG5nIjoyLjI0MTQxNjE2ODIyMTU3NjMsImxhdCI6NDEuMzg5NTQyNjg0NTYwM30sImciOjEwLjk3MDU2NjU1NDMwMDQ1NiwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [ ] Yes
- [ ] No
- [x] **Unsure** (let's discuss further)

## Case 3
A user has defined a unique color for each unique district in the data.  In this case, we want each category to be assigned to its defined color.

### Test

`color: ramp(buckets($n_distri,["Les Corts","Ciutat Vella","Sant Mart","Nou Barris","Eixample","Grcia","Horta-Guinard","Sarri-Sant Gervasi","Sant Andreu","Sants-Montjuc"]),[red,orange,yellow,green,blue,purple,sienna,fuchsia,coral,turquoise])`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcChidWNrZXRzKCRuX2Rpc3RyaSxbXCJMZXMgQ29ydHNcIixcIkNpdXRhdCBWZWxsYVwiLFwiU2FudCBNYXJ0XCIsXCJOb3UgQmFycmlzXCIsXCJFaXhhbXBsZVwiLFwiR3JjaWFcIixcIkhvcnRhLUd1aW5hcmRcIixcIlNhcnJpLVNhbnQgR2VydmFzaVwiLFwiU2FudCBBbmRyZXVcIixcIlNhbnRzLU1vbnRqdWNcIl0pLFtyZWQsb3JhbmdlLHllbGxvdyxncmVlbixibHVlLHB1cnBsZSxzaWVubmEsZnVjaHNpYSxjb3JhbCx0dXJxdW9pc2VdKSIsImYiOnsibG5nIjoyLjI0MTQxNjE2ODIyMTU3NjMsImxhdCI6NDEuMzg5NTQyNjg0NTYwM30sImciOjEwLjk3MDU2NjU1NDMwMDQ1NiwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [x] **Yes**
- [ ] No
- [ ] Unsure

## Case 4
A user has defined **all** categories in buckets, but has not defined enough colors. 

In this case, we will interpolate between a subset of the colors. The first category bucket will be assigned to the first color and the remaining buckets will be interpolated between the remaining colors.

### Test
`color: ramp(buckets($n_distri,["Les Corts","Ciutat Vella","Sant Mart","Nou Barris","Eixample","Grcia","Horta-Guinard","Sarri-Sant Gervasi","Sant Andreu","Sants-Montjuc"]),[red,blue,green,yellow])`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcChidWNrZXRzKCRuX2Rpc3RyaSxbXCJMZXMgQ29ydHNcIixcIkNpdXRhdCBWZWxsYVwiLFwiU2FudCBNYXJ0XCIsXCJOb3UgQmFycmlzXCIsXCJFaXhhbXBsZVwiLFwiR3JjaWFcIixcIkhvcnRhLUd1aW5hcmRcIixcIlNhcnJpLVNhbnQgR2VydmFzaVwiLFwiU2FudCBBbmRyZXVcIixcIlNhbnRzLU1vbnRqdWNcIl0pLFtyZWQsYmx1ZSxncmVlbix5ZWxsb3ddKSIsImYiOnsibG5nIjoyLjI0MTQxNjE2ODIyMTU3NjMsImxhdCI6NDEuMzg5NTQyNjg0NTYwM30sImciOjEwLjk3MDU2NjU1NDMwMDQ1NiwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

elenatorro commented:
Hi @makella, thank you! Those test cases are REALLY useful to have an example of each case ✨  

The one that you mention is the one we discussed in this thread. I was not sure how to solve it, but it seems that we agree with:

> Use the first n required colors and ignore the remaining ones

But my question is: what happens with "other" categories?
a) Should we use the last color in the array?
b) Should we use the `defaultColor`?

After thinking about it, I am not sure that we should avoid interpolation in this case, because what if the user's intention is to interpolate, and that is why she uses more colors than categories?

makella commented:
@elenatorro thinking through the question... will do some testing and get back to you with my thoughts!

makella commented:
@elenatorro after thinking about this a little bit, I think in this case (2b) that we should do the same thing that we are doing in **2a**. For category maps, that makes the most sense to me where if a user is defining buckets that they will expect to see the colors they defined used. If there are more colors, we can use the nth color for "others". I think that will also help people learn how this works.

I think this will also be a less common case. If, in the future we see that people want interpolation on this case, we can reconsider then.

How does that sound?

Regarding your question,
> After thinking about it, I am not sure that we should avoid interpolation in this case, because what if the user's intention is to interpolate, and that is why she uses more colors than categories?

In this case, if the user wants interpolation, they will figure out that they should add less colors like in Case 4.

Also, there will be the case where a user does `ramp($n_distri,[red,orange,blue])` which will send them to interpolation as well.
http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoiYmFycmlzX2JhcmNlbG9uYSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJjb2xvcjogcmFtcCgkbl9kaXN0cmksW3JlZCxvcmFuZ2UsYmx1ZV0pIiwiZiI6eyJsbmciOjIuMTk2OTE1NDg2ODM3NDk3LCJsYXQiOjQxLjM4NTA3MzU2MjIxNjk3NH0sImciOjEwLjk3MDU2NjU1NDMwMDQ1NiwiaCI6IkRhcmtNYXR0ZXIifQ==

Question:

Do you want me to go through these same scenarios for the CARTOColor schemes?

elenatorro commented:
Perfect, got it! 👍 

> Do you want me to go through these same scenarios for the CARTOColor schemes? 

Yes, please :) Thank you very much!

makella commented:
Hi @elenatorro I realized that we also need to test all the cases with custom sequential schemes and sequential and diverging CARTOColor schemes for numeric data!

Have we decided if we are separating the rules/defaults for `string` and `numeric?  Thinking through, I see cases where there are certain rules that apply only to `string` and others only to `numeric`. (For example, the default "others" color among others). 

I'm not sure if we've discussed that in too much detail so before proceeding through those test cases, we should make a decision about that part!

----

In the meantime, below are the tests for **Categories/CARTOColors**

For this test, I'm using a polygon dataset of South American countries with a total of **13** unique values using the CARTOColor qualitative scheme `bold`.

`'PER','URY','BOL','BRA','SUR','VEN','CHL','GB1','PRY','ARG','COL','ECU','GUY'`

After going through all of them, there are a couple of things that are different when you use our color schemes or a custom one.  Specifically here **2a, 3, and 4**.

## Case 1a
A user wants to color four countries with a unique color from `bold`. In this case, all other features will be colored with the color scheme's neutral `other` gray.

### Test
`color: ramp(buckets($sov_a3,['PER','URY','BOL','BRA']),bold)`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoYnVja2V0cygkc292X2EzLFsnUEVSJywnVVJZJywnQk9MJywnQlJBJ10pLGJvbGQpXG5cbiIsImYiOnsibG5nIjotNDUuMzA5MDQxMTY0NTcwMDksImxhdCI6LTI4LjI2NDcwMzMyNjc2Mjc3OH0sImciOjIuMTg1MjYxNzY0MzIxMjkwNCwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

## Case 1b
A user has defined an equal number of category buckets to colors (**11**) in the CARTOColor scheme, `bold`. In this case, we color each bucket with the 11 colors from `bold` and color everything else with `bold` default "other" color.

### Test
`color: ramp(buckets($sov_a3,['PER','URY','BOL','BRA','SUR','VEN','CHL','GB1','PRY','ARG','COL']),bold)`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoYnVja2V0cygkc292X2EzLFsnUEVSJywnVVJZJywnQk9MJywnQlJBJywnU1VSJywnVkVOJywnQ0hMJywnR0IxJywnUFJZJywnQVJHJywnQ09MJ10pLGJvbGQpXG5cbiIsImYiOnsibG5nIjotNDUuMzA5MDQxMTY0NTcwMDksImxhdCI6LTI4LjI2NDcwMzMyNjc2Mjc3OH0sImciOjIuMTg1MjYxNzY0MzIxMjkwNCwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

## Case 2a

### **_Not applicable to CARTOColors qualitative/buckets. Since you can't add an additional color._**

### Test
`color: ramp(buckets($sov_a3,['PER','URY','BOL','BRA','SUR','VEN','CHL','GB1','PRY','ARG','COL','ECU','GUY']),bold)`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoYnVja2V0cygkc292X2EzLFsnUEVSJywnVVJZJywnQk9MJywnQlJBJywnU1VSJywnVkVOJywnQ0hMJywnR0IxJywnUFJZJywnQVJHJywnQ09MJywnRUNVJywnR1VZJ10pLGJvbGQpXG5cbiIsImYiOnsibG5nIjotNDUuMzA5MDQxMTY0NTcwMDksImxhdCI6LTI4LjI2NDcwMzMyNjc2Mjc3OH0sImciOjIuMTg1MjYxNzY0MzIxMjkwNCwiaCI6IkRhcmtNYXR0ZXIifQ==

## Case 2b

A user has defined a subset of country buckets (**8**) using `bold`. In this case, we will use the first 8 colors in `bold` and color everything else with `bold` other color.

### Test
`color: ramp(buckets($sov_a3,['PER','URY','BOL','BRA','SUR','VEN','CHL','GB1']),bold)`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoYnVja2V0cygkc292X2EzLFsnUEVSJywnVVJZJywnQk9MJywnQlJBJywnU1VSJywnVkVOJywnQ0hMJywnR0IxJ10pLGJvbGQpXG5cbiIsImYiOnsibG5nIjotNDUuMzA5MDQxMTY0NTcwMDksImxhdCI6LTI4LjI2NDcwMzMyNjc2Mjc3OH0sImciOjIuMTg1MjYxNzY0MzIxMjkwNCwiaCI6IkRhcmtNYXR0ZXIifQ== 

### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

## Case 3

A user creates buckets for each unique country in the data (**13**) using `bold`. In this case, we will **interpolate** because we don't have 13 colors available in `bold`.

### Test

`color: ramp(buckets($sov_a3,['PER','URY','BOL','BRA','SUR','VEN','CHL','GB1','PRY','ARG','COL','ECU']),bold)`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoYnVja2V0cygkc292X2EzLFsnUEVSJywnVVJZJywnQk9MJywnQlJBJywnU1VSJywnVkVOJywnQ0hMJywnR0IxJywnUFJZJywnQVJHJywnQ09MJywnRUNVJ10pLGJvbGQpXG5cbiIsImYiOnsibG5nIjotNDUuMzA5MDQxMTY0NTcwMDksImxhdCI6LTI4LjI2NDcwMzMyNjc2Mjc3OH0sImciOjIuMTg1MjYxNzY0MzIxMjkwNCwiaCI6IkRhcmtNYXR0ZXIifQ==

### Passed?
- [X] **Yes**
- [ ] No
- [ ] Unsure

## Case 4
A user has defined **all** categories in buckets, but has not defined enough colors.  In this case, we will interpolate between a subset of the colors. 

### **_For CARTOColor qualitative schemes, this would be similar to Case 3 since the colors aren't being defined by the user. Therefore, if there are more categories than colors, we will go to interpolation._**

### Test
`color: ramp(buckets($sov_a3,['PER','URY','BOL','BRA','SUR','VEN','CHL','GB1','PRY','ARG','COL','ECU','GUY']),bold)`

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoYnVja2V0cygkc292X2EzLFsnUEVSJywnVVJZJywnQk9MJywnQlJBJywnU1VSJywnVkVOJywnQ0hMJywnR0IxJywnUFJZJywnQVJHJywnQ09MJywnRUNVJywnR1VZJ10pLGJvbGQpXG5cbiIsImYiOnsibG5nIjotNDUuMzA5MDQxMTY0NTcwMDksImxhdCI6LTI4LjI2NDcwMzMyNjc2Mjc3OH0sImciOjIuMTg1MjYxNzY0MzIxMjkwNCwiaCI6IkRhcmtNYXR0ZXIifQ==

elenatorro commented:
Hi @makella, 

Totally 👍  I agree we should discuss test cases for numerical values, I will take a look and we can talk later about it.

makella commented:
@elenatorro ok, please do! My calendar is up-to-date if you want to schedule something for us.

elenatorro commented:
Hi again! I have updated the issue and checked the tests you marked as "yes". The **Case 2b** is still ongoing, and I have added the test descriptions for numeric properties (**and properties are numbers**).

makella commented:
To test:

http://127.0.0.1:8080/examples/editor/index.html#eyJhIjoibmVfNTBtX2FkbWluXzBfY291bnRyaWVzX2xha2VzX2NvcHkiLCJiIjoiIiwiYyI6Im1hbWF0YWFrZWxsYSIsImQiOiJodHRwczovL3t1c2VyfS5jYXJ0by5jb20iLCJlIjoiY29sb3I6IHJhbXAoJHNvdl9hMyxbcmVkLHB1cnBsZSxvcmFuZ2VdKVxuc3Ryb2tlV2lkdGg6IDBcbiIsImYiOnsibG5nIjotNDguMzE5ODE0NTA0MTI5MTcsImxhdCI6LTcuNDY5MzQzMzU3MjkwMjk5fSwiZyI6MywiaCI6IkRhcmtNYXR0ZXIifQ==

elenatorro commented on src/renderer/viz/expressions/buckets.js:
```diff
@@ -90,41 +92,55 @@ export default class Buckets extends BaseExpression {
         this.numCategories = list.elems.length + 1;
         this.list = list;
         this.type = 'category';
+        this.isBuckets = true;
```
Yes, I am actually using now `input.isA(Bucket)` so this is no longer needed.

elenatorro commented on src/renderer/viz/expressions.js:
```diff
@@ -191,8 +191,8 @@ import { Desc } from './expressions/ordering';
-import { palettes } from './expressions/color/palettes';
-import { Reverse } from './expressions/color/palettes';
+import palettes from './expressions/color/palettes';
+import PaletteReverse from './expressions/color/palettes/PaletteReverse';
```
Is not that long :)

davidmanzanares replied:
You'll need to convince others :slightly_smiling_face:

IagoLast replied:
does it breaks the API?

elenatorro replied:
no, it has been tested

davidmanzanares replied:
I think it's quite awkward to have a different internal name than the exported name: `+export const reverse = (...args) => new PaletteReverse(...args);` This is the first time an expression class doesn't share its name with the exported name

IagoLast replied:
yep, I think we should be consistent with this, in any case maybe in the future we overload the method allowing different parameters types but palettes ant in this case `palleteReverse`would be a bad name :)

elenatorro replied:
I agree. The truth is that `reverse` for me is really generic, you can actually "reverse" other values, not only a palette. That is why IMO it should be `paletteReverse` o `reversePalette`. But I think we should not change now the expression name in this PR. I can name the class Reverse and we will discuss this in the future.

davidmanzanares replied:
Yes, that's a good reason to make it more generic in the future, which was the original idea. Could you open a ticket to track that?

elenatorro commented on src/renderer/viz/expressions/buckets.js:
```diff
@@ -90,41 +92,55 @@ export default class Buckets extends BaseExpression {
+
     _compile(metadata) {
         super._compile(metadata);
-        this.othersBucket = this.input.type == 'category';
```
Yep, also unnecessary

davidmanzanares replied:
I don't understand your reply :confused:

elenatorro replied:
Ups, I read it wrong, I thought you were asking me this property to be removed. After reading `ramp.js`, do you still think we need it?

davidmanzanares replied:
Is not needed after the refactor in `ramp.js`, yet my point is that this is a regression from the quality of the code point of view. Instead of having a global property for every expression and being generic this PR has checks like `instanceof Top` and `instaceof Buckets`

elenatorro replied:
I agree. Since we are going to change that in a short term, I would refactor that in a different PR because this code will change.

davidmanzanares replied:
I'll buy it

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -61,26 +87,26 @@ export default class Ramp extends BaseExpression {
-        } else {
-            this.type = 'color';
-        }
+        this.inlineMaker = () => undefined;
```
It breaks otherwise, I will take a look.

elenatorro replied:
fixed

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -97,28 +123,63 @@ export default class Ramp extends BaseExpression {
+        const pixelValues = this._computeTextureIfNeeded();
+        const input = this.input.eval(feature);
+
+        if (input !== -1) {
```
Depending on the input, it can be looking for an element in an array. If this element does not exist, it returns -1.

davidmanzanares replied:
Who returns -1?

elenatorro replied:
fixed

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -97,28 +123,63 @@ export default class Ramp extends BaseExpression {
+    }
+    
+    _getColorValue(pixelValues, m) {
+        const index = Math.round(m * MAX_BYTE_VALUE);
```
I do not think so because otherwise there will be indexes out of the array. Actually, the bad thing here is the naming... probably it should be `COLOR_ARRAY_LENGTH - 1` (which in this case is the same as MAX_BYTE_VALUE, but that is a coincidence in this case).

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -97,28 +123,63 @@ export default class Ramp extends BaseExpression {
+        checkType('ramp', 'input', 0, Object.values(inputTypes), this.input);
+        
+        if (this.palette.type === paletteTypes.SPRITE) {
+            checkType('ramp', 'input', 0, inputTypes.CATEGORY, this.input);
```
is this good or bad? 😂

davidmanzanares replied:
hahaha I guess good, but I'll go cry in a corner

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -131,116 +192,120 @@ export default class Ramp extends BaseExpression {
+
+        const othersColor = new NamedColor('gray');
+        return palette.type === paletteTypes.PALETTE
+            ? _getColorsFromPaletteType(input, palette, this.maxKey, othersColor.eval())
```
totally agree!! How would you rename it?

davidmanzanares replied:
I no longer know what it is :joy:

elenatorro replied:
ok, I will take a look.

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -131,116 +192,120 @@ export default class Ramp extends BaseExpression {
+                : this._computeTexture();
         }
+
+        return [];
```
This function should always return an array.

davidmanzanares replied:
This smells a little bit... in which cases does this happen?

elenatorro replied:
The methods `_computeTextureColor` and `_computeTexture` (which are going to be renamed)  always return an array. That is why if `this._texCategories !== this.input.numCategories` is false, I return `[]` to make this method always to return an array.

elenatorro replied:
Fixed ✨

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -131,116 +192,120 @@ export default class Ramp extends BaseExpression {
+        const floats = this.palette.floats;
+
+        for (let i = 0; i < COLOR_ARRAY_LENGTH; i++) {
+            const vlowRaw = floats[Math.floor(i / MAX_BYTE_VALUE * (floats.length - 1))];
```
Same as above, I am not using constants properly 👍

davidmanzanares replied:
We may want to change the color array to something different than `256`, that way we'll know if things are ok, or if they :boom:

elenatorro replied:
👍 Right, I will change it to `COLOR_ARRAY_LENGTH - 1` where needed.

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -265,26 +333,119 @@ export default class Ramp extends BaseExpression {
+        colors = _getSubPalettes(palette, numCategories);
+        othersColor = colors[numCategories];
+    }
+
```
You are totally right, I will refactor it.

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -265,26 +333,119 @@ export default class Ramp extends BaseExpression {
+    }
+
+    if (palette.isQualitative()) {
+        if (input.isA(Top)) {
```
That was one of the questions that I had! And yes, that is right! I should do the same for both `top` and `buckets` expressions.

davidmanzanares replied:
I think this will be easier with one of the first comments I made in `buckets` regarding the removal of `othersBucket`

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -265,26 +333,119 @@ export default class Ramp extends BaseExpression {
+function _getColorsFromColorArrayTypeNumeric(numCategories, colors) {
+    let otherColor;
+    
+    if (numCategories < colors.length) {
```
Yes, this might be wrong, I will check it.

elenatorro replied:
After looking at it, it is because it can be categories with numeric values.

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -265,26 +333,119 @@ export default class Ramp extends BaseExpression {
+    return colors;
+}
+
+function _addOtherColorToColors (colors, otherColor) {
```
I am used to naming private-private (:joy:) functions with underscore.

IagoLast replied:
I think underscore is a safety check to reveal the intention of not-exporting a function

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -265,26 +333,119 @@ export default class Ramp extends BaseExpression {
+            colorArray.push(colors[i]);
+        } else if (i === numCategories) {
+            colorArray.push(othersColor);
+        }
```
In this case there is interpolation, so this is in order to avoid (showing) interpolation.

davidmanzanares replied:
:confused:

elenatorro commented:
Thanks for this awesome review @davidmanzanares 👌

elenatorro commented on src/renderer/viz/colorspaces.js:
```diff
@@ -7,6 +7,31 @@ export function cielabToSRGB(cielab) {
 }
 
+export function interpolate(low, high, m) {
+    const cielabLow = sRGBToCielab({
```
Why not?

davidmanzanares replied:
I initially used `low`/`high` in the context of having a floating-point index in the [0,1] range and mapping that to two values in the array: the low value and the high value. For me is strange to read low/high here, something even simpler like a/b is better from my point of view

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -97,28 +123,63 @@ export default class Ramp extends BaseExpression {
-        if (this.palette.type != 'number-array') {
-            super.eval(o);
+    eval(feature) {
+        const pixelValues = this._computeTextureIfNeeded();
```
fixed

elenatorro commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -97,28 +123,63 @@ export default class Ramp extends BaseExpression {
+                : this._getColorValue(pixelValues, m);
+            
+            this._texCategories = null;
+            this._GLtexCategories = null;
```
Fixed, you were right! It was unnecessary at all :)

davidmanzanares commented on src/renderer/viz/expressions/ramp.js:
```diff
@@ -1,7 +1,33 @@
+import Linear from './linear';
+import Top from './top';
+
+const paletteTypes = {
```
ping :slightly_smiling_face:

elenatorro replied:
I think it is more readable, at list for this long class, to have a place to look when you want to know which are the different palette types available. Is not a Java-JavaScript-OtherLanguage issue, but it is a good practice to avoid both strings and magic numbers when possible. I also thing this constants should be in a separate file along with other constants so we can have them available when needed, but that is another discussion 🙈

davidmanzanares commented on test/unit/renderer/viz/expressions/basic/category.test.js:
```diff
@@ -22,4 +22,26 @@ describe('src/renderer/viz/expressions/basic/category', () => {
     });
 
+    describe('.eval', () => {
+        const fakeMetadata = {
```
:heart:

davidmanzanares commented on test/unit/renderer/viz/expressions/basic/category.test.js:
```diff
@@ -22,4 +22,26 @@ describe('src/renderer/viz/expressions/basic/category', () => {
+            columns: [{
+                type: 'category',
+                name: 'category',
+                categoryNames: ['cat0', 'cat1', 'cat2']
```
I think `categoryNames` is the old Metadata API

davidmanzanares commented on test/unit/renderer/viz/expressions/buckets.test.js:
```diff
@@ -14,6 +16,227 @@ describe('src/renderer/viz/expressions/buckets', () => {
+                    city: {
+                        type: 'category',
+                        categories: [
+                            { city: 'Murcia' }, 
```
the API for this I think is `categories: [{name: 'Murcia'}...` without repeating the property name `city`

elenatorro replied:
Nope, after testing it, it needs to be the name of the property, right?

davidmanzanares replied:
I don't think so :sweat_smile: https://github.com/CartoDB/carto-vl/blob/master/src/renderer/Metadata.js#L22 and https://github.com/CartoDB/carto-vl/blob/master/test/unit/renderer/viz/expressions/belongs.test.js#L10 Could you recheck the implementation too?

elenatorro replied:
Ok, I will change that :) I may need to change some of the specs.

davidmanzanares commented on test/unit/renderer/viz/expressions/ramp.test.js:
```diff
@@ -1,5 +1,11 @@
+import { hexToRgb } from '../../../../../src/renderer/viz/expressions/utils';
+import Metadata from '../../../../../src/renderer/Metadata';
+
+const DEFAULT_COLOR = s.namedColor('gray');
```
Couldn't we get the default color from the ramp? It could be interesting for end-users to be able to do that to make legends.

elenatorro replied:
I think we should address that in a different issue.

davidmanzanares replied:
:+1:

davidmanzanares commented on test/unit/renderer/viz/expressions/basic/category.test.js:
```diff
@@ -23,18 +24,18 @@ describe('src/renderer/viz/expressions/basic/category', () => {
+                category: {
+                    type: 'category',
+                    categories: [
+                        { category: 'cat0' }, 
```
`name`, not `category` :slightly_smiling_face:

elenatorro commented:
Ready to merge? cc @IagoLast @davidmanzanares @jgoizueta

IagoLast commented:
Example of how to taclke an issue! Good description and summary, followed by a depth discussion with all the team about posible edge-cases and crazy situations, unit-tests describing most of the expected cases, discovering and fixing bugs... congrats 🥇

makella commented:
yes!! great work all around!! also, since we have already gone into so many tests with this, we are well equipped to have a guide! :) next week, i'll get that sorted out better.
